codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/utils/CaseUtils.kt

```diff
@@ -8,32 +8,109 @@ package software.amazon.smithy.kotlin.codegen.utils
+    val allLowerCase = lowercase() == this
+    forEachIndexed { index, nextChar ->
+        val peek = getOrNull(index + 1)
+        val doublePeek = getOrNull(index + 2)
+        val completeWordInProgress = completeWords.any {
+            (currentWord + substring(index)).lowercase().startsWith(it)
+        } && !completeWords.contains(currentWord.lowercase())
 
-    // add a space between camelCased words
-    result = result.split(Regex("(?<=[a-z])(?=[A-Z]([a-zA-Z]|[0-9]))")).joinToString(separator = " ") // AcmSuccess -> // "Acm Success"
+        when {
+            // [C] in these docs indicates the value of nextCharacter
+            // A[_]B
+            !nextChar.isLetterOrDigit() -> emit(nextChar)
 
-    // add a space after acronyms
-    result = result.replace(Regex("([A-Z]+)([A-Z][a-z])"), "$1 $2") // "ACMSuccess" -> "ACM Success"
+            // If we have no letters so far, push the next letter (we already know it's a letter or digit)
+            currentWord.isEmpty() -> currentWord += nextChar.toString()
 
-    // add space after a number in the middle of a word
-    result = result.replace(Regex("([0-9])([a-zA-Z])"), "$1 $2") // "s3ec2" -> "s3 ec2"
+            // Abc[D]ef or Ab2[D]ef
+            !completeWordInProgress && loweredFollowedByUpper(currentWord, nextChar) -> emit(nextChar)
 
-    // remove extra spaces - multiple consecutive ones or those and the beginning/end of words
-    result = result.replace(Regex("\\s+"), " ") // "Foo  Bar" -> "Foo Bar"
-        .trim() // " Foo " -> "Foo"
+            // s3[k]ey
+            !completeWordInProgress && allLowerCase && digitFollowedByLower(currentWord, nextChar) -> emit(nextChar)
 
-    return result.split(" ")
+            // DB[P]roxy, or `IAM[U]ser` but not AC[L]s
+            endOfAcronym(currentWord, nextChar, peek, doublePeek) -> emit(nextChar)
+
+            // emit complete words
+            !completeWordInProgress && completeWords.contains(currentWord.lowercase()) -> emit(nextChar)
+
+            // If we haven't found a word boundary, push it and keep going
+            else -> currentWord += nextChar.toString()
+        }
+    }
+    if (currentWord.isNotEmpty()) {
+        out += currentWord
+    }
+
+    return out
 }
 
+/**
+ * Handle cases like `DB[P]roxy`, `ARN[S]upport`, `AC[L]s`
+ */
+private fun endOfAcronym(current: String, nextChar: Char, peek: Char?, doublePeek: Char?): Boolean {
+    if (!current.last().isUpperCase()) {
+        // Not an acronym in progress
+        return false
+    }
+    if (!nextChar.isUpperCase()) {
+        // We aren't at the next word yet
+        return false
+    }
+
+    if (peek?.isLowerCase() != true) {
+        return false
+    }
+
+    // Skip cases like `AR[N]s`, `AC[L]s` but not `IAM[U]ser`
+    if (peek == 's' && (doublePeek == null || !doublePeek.isLowerCase())) {
+        return false
+    }
+
+    // Skip cases like `DynamoD[B]v2`
+    return !(peek == 'v' && doublePeek?.isDigit() == true)
```

Comment:
**Correctness**/**Question**: Why do we want to skip cases like this? It seems to me that "DynamoDBv2" in this case represents the words `["Dynamo", "DB", "v2"]` not `["Dynamo", "DBv2"]`. I think "IPv4" and "IPv6" are exceptions because they're actually whole initialisms but I think _generally_ "v*n*" is the start of a new word. Several test expectations below changed for the worse IMHO (e.g., `"TlsV1_2"` → `"Tlsv1_2"`, `"MsChapV1"` → `"MsChapv1"`, etc.).

Reply:
Changing this results in:

**Member Names**
```
DynamoDBv2Action,dynamoDbv2Action => dynamoDBv2Action (expected dynamoDbv2Action)
dynamoDBv2,dynamoDbv2 => dynamoDBv2 (expected dynamoDbv2)
```

**Enum Names**
```
expected 'MsChapv1' != actual 'MsChaPv1' for input: MS-CHAPv1
expected 'Tlsv1' != actual 'TlSv1' for input: TLSv1
expected 'Tlsv1_2' != actual 'TlSv1_2' for input: TLSv1.2
org.opentest4j.AssertionFailedError: expected 'MsChapv1' != actual 'MsChaPv1' for input: MS-CHAPv1
expected 'Tlsv1' != actual 'TlSv1' for input: TLSv1
expected 'Tlsv1_2' != actual 'TlSv1_2' for input: TLSv1.2
```

**Client Names**

```
SESv2,Sesv2 => SeSv2 (expected Sesv2)
```

Reply:
Right, turns out I didn't fully understand the `endOfAcronym` function. Since it's working on determining whether `current` is the end, I think we need to add a new special handling case between `// Not an acronym in progress` and `// We aren't at the next word yet`:

```kotlin
if (nextChar == 'v' && peek?.isDigit() == true) {
    // Handle cases like `DynamoDB[v]2`
    return true
}
```

This yields `dynamoDbV2`, `MsChapV1`, `TlsV1_2`, etc.

<details>
  <summary>Full function code</summary>
  
  ```kotlin
  private fun endOfAcronym(current: String, nextChar: Char, peek: Char?, doublePeek: Char?): Boolean {
      if (!current.last().isUpperCase()) {
          // Not an acronym in progress
          return false
      }
  
      if (nextChar == 'v' && peek?.isDigit() == true) {
          // Handle cases like `DynamoDB[v]2`
          return true
      }
  
      if (!nextChar.isUpperCase()) {
          // We aren't at the next word yet
          return false
      }
  
      if (peek?.isLowerCase() != true) {
          return false
      }
  
      // Skip cases like `AR[N]s`, `AC[L]s` but not `IAM[U]ser`
      if (peek == 's' && doublePeek?.isLowerCase() != true) {
          return false
      }
  
      // Skip cases like `DynamoD[B]v2`
      return !(peek == 'v' && doublePeek?.isDigit() == true)
  }
  ```
</details>